### PR TITLE
Blog App: validations and model specs

### PR DESCRIPTION
**Pull request by quavo19**
### In this milestone I;

- Added validations For the User model
 - Added validations For the  Post model
 - Added unit specs for all of my models, methods, and validations.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  it 'is invalid with a negative posts_counter' do
+    user.posts_counter = -1
+    expect(user).to_not be_valid
+  end
+
+  it 'is valid with a zero posts_counter' do
+    user.posts_counter = 0
+    expect(user).to be_valid
+  end
```
Comment (danmainah) [inline spec/models/user_spec.rb:23]: - [ ] @quavo19 Great work adding tests to user model validation, however, you forgot to add a test for the ` recent_posts` method. This is a project requirement as highlighted below:

image

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    it { should validate_numericality_of(:comments_counter).is_greater_than_or_equal_to(0) }
+    it { should validate_numericality_of(:likes_counter).is_greater_than_or_equal_to(0) }
+  end
+
+  describe 'callbacks' do
+    describe 'after_save' do
+      it 'updates posts_counter of author' do
+        expect { subject.save }.to change { user.reload.posts_counter }.from(0).to(1)
+      end
+    end
```
Comment (danmainah) [inline spec/models/post_spec.rb:23]: - [ ] @quavo19 great work on adding tests for the post validations and methods however, you forgot to add tests for the `recent_comments` method.

File: spec/models/like_spec.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    post = user.posts.build(title: 'Post 1', comments_counter: 0, likes_counter: 0)
+    post.author_id = user.id
+    # Create a like
+    post.save!
+    like = post.likes.build(author_id: user.id)
+    like.post_id = post.id
+    # Increment likes_counter
+    like.save!
+    expect(Post.find(1).likes_counter).to eq(1)
+  end
```
Comment (danmainah) [inline spec/models/like_spec.rb]: - [x] @quavo19 Great work on your tests, however, this test is failing when I run tests as shown below, I would recommend  you check how you are creating the user post and like
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @quavo19 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️ Github flow was correctly used
✔️ Professional README file.

# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/models/user_spec.rb (hunk 1 above)
Comment (quavo19) [reply spec/models/user_spec.rb:23]: hello, @danmainah those methods are private methods so they can't be accessed outside the function. that's why they can't be tested.
i can go ahead and remove `Private` from my files and submit if that is what you would like me to do

**Review by alexpuente0 — APPROVED** [review]
Hi @quavo19 , 



Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- Great job with the tests, everything clears out and you implemented the suggestions from the last reviewer.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**